### PR TITLE
add ecdsa configuration options

### PR DESCRIPTION
**Pull request by pszal**
Add configuration options for ECDSA key generation.

### CODE REVIEW

File: attest/application_key.go [hunk 1]
```diff
@@ -56,9 +56,27 @@ func (s *signer) Public() crypto.PublicKey {
...
 }
 
-// KeyConfig encapsulates parameters for minting keys. This type is defined
-// now (despite being empty) for future interface compatibility.
+// Algorithm indicates an asymmetric algorithm to be used.
+type Algorithm string
+
+// Algorithm types supported.
+const (
+	EC  Algorithm = "EC"
```
Comment (ericchiang) [inline attest/application_key.go:64]: nit: I think this should be "ECDSA"

File: attest/application_key.go [hunk 2]
```diff
@@ -56,9 +56,27 @@ func (s *signer) Public() crypto.PublicKey {
...
+	RSA Algorithm = "RSA"
```
Comment (ericchiang) [inline attest/application_key.go:65]: If we export this variable can you add a comment explaining that it's not supported yet?

File: attest/application_key.go [hunk 3]
```diff
@@ -56,9 +56,27 @@ func (s *signer) Public() crypto.PublicKey {
...
+)
+
+// KeyConfig encapsulates parameters for minting keys.
 type KeyConfig struct {
+	// Algorithm to be used, either RSA or EC.
+	Algorithm Algorithm
+	// Size is used to specify the bit size of the key.
```
Comment (ericchiang) [inline attest/application_key.go:72]: "// Size is used to specify the bit size of the key or elliptic curve. For example, '256' is used to specify curve P-256."

File: attest/application_key.go [hunk 4]
```diff
@@ -56,9 +56,27 @@ func (s *signer) Public() crypto.PublicKey {
...
+	Size int
+}
+
+// DefaultConfig is used when no other configuration is specified.
+var DefaultConfig = KeyConfig{
```
Comment (ericchiang) [inline attest/application_key.go:77]: No real need to export this? You can already specify the DefaultConfig by using nill or an empty key config?

File: attest/tpm.go [hunk 5]
```diff
@@ -354,8 +348,12 @@ func (t *TPM) NewAK(opts *AKConfig) (*AK, error) {
 	return t.tpm.newAK(opts)
 }
 
-// NewKey creates an application key certified by the attestation key.
+// NewKey creates an application key certified by the attestation key. If opts is nil
+// then DefaultConfig is used.
 func (t *TPM) NewKey(ak *AK, opts *KeyConfig) (*Key, error) {
+	if opts == nil {
```
Comment (ericchiang) [inline attest/tpm.go:354]: This breaks the following code?

```
t.NewKey(ak, &attest.KeyConfig{})
```

Probably do something like

```
var defaultConfig = &KeyConfig{
  // ...
}

// ...
if opts == nil {
    opts = &KeyConfig{}
}
if opts.Algorithm == "" && opts.Size == 0 {
    opts = defaultConfig
}
```

File: attest/wrapped_tpm20.go [hunk 6]
```diff
@@ -206,6 +201,68 @@ func (t *wrappedTPM20) newKey(ak *AK, opts *KeyConfig) (*Key, error) {
...
+	}
+
+	return srk, blob, pub, creationData, err
+}
+
+func templateFromConfig(opts *KeyConfig) (tpm2.Public, error) {
+	var tmpl tpm2.Public
+	switch opts.Algorithm {
+	case RSA:
+		return tmpl, fmt.Errorf("not implemented")
```
Comment (ericchiang) [inline attest/wrapped_tpm20.go:227]: nit: more specific error message

```
return tmpl, fmt.Errorf("RSA keys are not implemented")
```

File: attest/application_key.go (hunk 1 above)
Comment (pszal) [reply attest/application_key.go]: I wanted to keep it consistent with certtostore but indeed ECDSA is more precise. Done.

File: attest/application_key.go (hunk 2 above)
Comment (pszal) [reply attest/application_key.go:66]: Done. I plan to do an RSA follow-up.

File: attest/application_key.go (hunk 3 above)
Comment (pszal) [reply attest/application_key.go]: Done.

File: attest/application_key.go (hunk 4 above)
Comment (pszal) [reply attest/application_key.go]: Done.

File: attest/tpm.go (hunk 5 above)
Comment (pszal) [reply attest/tpm.go:354]: Done.

File: attest/wrapped_tpm20.go (hunk 6 above)
Comment (pszal) [reply attest/wrapped_tpm20.go]: Done.

File: attest/application_key.go [hunk 7]
```diff
@@ -56,9 +56,29 @@ func (s *signer) Public() crypto.PublicKey {
...
+	// TODO(szp): RSA is not supported yet
+	RSA Algorithm = "RSA"
+)
+
+// KeyConfig encapsulates parameters for minting keys.
 type KeyConfig struct {
+	// Algorithm to be used, either RSA or ECDSA.
+	Algorithm Algorithm
+	// Size is used to specify the bit size of the key or elliptic curve. For
+	// example, '256' is used to specify curve P-256."
```
Comment (ericchiang) [inline attest/application_key.go:74]: nit: looks like a trailing '"' snuck in here. can you remove it?

**Review by ericchiang — APPROVED** [review]
one nit, please squash, then :shipit:

**Pull request merged**